core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1188,119 +1133,201 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
+        controllerContext.partitionReplicaAssignment.++=(addedPartitionReplicaAssignment)
+        info("New topics: [%s], deleted topics: [%s], new partition replica assignment [%s]".format(newTopics,
+          deletedTopics, addedPartitionReplicaAssignment))
+        if (newTopics.nonEmpty)
+          onNewTopicCreation(newTopics, addedPartitionReplicaAssignment.keySet)
+      } catch {
+        case e: Throwable => error("Error while handling new topic", e)
+      }
+    }
+  }
+
+  case class PartitionModifications(topic: String) extends ControllerEvent {
+    override def process(): Unit = {
+      if (!isActive) return
+      try {
+        val partitionReplicaAssignment = zkUtils.getReplicaAssignmentForTopics(List(topic))
+        val partitionsToBeAdded = partitionReplicaAssignment.filter(p =>
+          !controllerContext.partitionReplicaAssignment.contains(p._1))
+        if(topicDeletionManager.isTopicQueuedUpForDeletion(topic))
+          error("Skipping adding partitions %s for topic %s since it is currently being deleted"
+            .format(partitionsToBeAdded.map(_._1.partition).mkString(","), topic))
+        else {
+          if (partitionsToBeAdded.nonEmpty) {
+            info("New partitions to be added %s".format(partitionsToBeAdded))
+            controllerContext.partitionReplicaAssignment.++=(partitionsToBeAdded)
+            onNewPartitionCreation(partitionsToBeAdded.keySet)
+          }
         }
+      } catch {
+        case e: Throwable => error("Error while handling add partitions for topic " + topic, e)
       }
     }
   }
 
-  def doHandleDataDeleted(dataPath: String) {}
-}
+  case class TopicDeletion(var topicsToBeDeleted: Set[String]) extends ControllerEvent {
+    override def process(): Unit = {
+      if (!isActive) return
+      debug("Delete topics listener fired for topics %s to be deleted".format(topicsToBeDeleted.mkString(",")))
+      val nonExistentTopics = topicsToBeDeleted -- controllerContext.allTopics
+      if (nonExistentTopics.nonEmpty) {
+        warn("Ignoring request to delete non-existing topics " + nonExistentTopics.mkString(","))
+        nonExistentTopics.foreach(topic => zkUtils.deletePathRecursive(getDeleteTopicPath(topic)))
+      }
+      topicsToBeDeleted --= nonExistentTopics
+      if (config.deleteTopicEnable) {
+        if (topicsToBeDeleted.nonEmpty) {
+          info("Starting topic deletion for topics " + topicsToBeDeleted.mkString(","))
+          // mark topic ineligible for deletion if other state changes are in progress
+          topicsToBeDeleted.foreach { topic =>
+            val preferredReplicaElectionInProgress =
+              controllerContext.partitionsUndergoingPreferredReplicaElection.map(_.topic).contains(topic)
+            val partitionReassignmentInProgress =
+              controllerContext.partitionsBeingReassigned.keySet.map(_.topic).contains(topic)
+            if (preferredReplicaElectionInProgress || partitionReassignmentInProgress)
```

Comment:
Hmm, is there a particular reason that you want to do the simplification in a separate patch? Sometimes we defer cleanups to avoid the rebasing overhead. However, in this case, no one else is touching the controller code. So this shouldn't be an issue. Also, could you preserve the commit history when updating the patch? That will make it easier to review the delta changes.